### PR TITLE
Eradicate coverity issues in ApplicationWindow.cpp

### PR DESCRIPTION
**Pull request by eXeC64**
This pull request is for trac ticket: [#11094](http://trac.mantidproject.org/mantid/ticket/11094)

There are almost 100 Coverity defects in ApplicationWindow.cpp and this pull request takes care of all of them.

I started by having a commit per issue (or two), for ease of code review but it became rather cumbersome. Instead I started grouping changes by their function.

### CODE REVIEW

**Comment by FedeMPouzols** [comment]
This is what I call a brave effort :)
MantidPlot was so much in need of this. I'll review it but all looks fine at first glimpse.

The build failed because of a random test failure on rhel6. I've seen that one before (TEST-AlgorithmsTest.RemoveExpDecayTest) some days ago. Not sure what's randomly wrong with it.

**Comment by FedeMPouzols** [comment]
All looks good to me. I just have a few comments that I'm writing inline. Let me know if those are worth fixing and I'll merge the branch.

Oh, in principle this looked ready to be merged in, but it seems that a merge conflict has popped up?

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 1]
```diff
@@ -8647,10 +8817,14 @@ void ApplicationWindow::undo()
...
 
-  if (qobject_cast<Note*>(w))
-    dynamic_cast<Note*>(w)->editor()->undo();
-  else if (qobject_cast<Matrix*>(w)){
-    QUndoStack *stack = (dynamic_cast<Matrix*>(w))->undoStack();
+  auto note = dynamic_cast<Note*>(w);
+  auto matrix = dynamic_cast<Matrix*>(w);
+
+  if(note)
+    note->editor()->undo();
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: Maybe `if(note && note->editor())` for better safety? There are a couple of similar points some 300 lines down...

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 2]
```diff
@@ -8665,10 +8839,14 @@ void ApplicationWindow::redo()
...
 
-  if (qobject_cast<Note*>(w))
-    dynamic_cast<Note*>(w)->editor()->redo();
-  else if (qobject_cast<Matrix*>(w)){
-    QUndoStack *stack = (dynamic_cast<Matrix*>(w))->undoStack();
+  auto note = dynamic_cast<Note*>(w);
+  auto matrix = dynamic_cast<Matrix*>(w);
+
+  if(note)
+    note->editor()->redo();
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: Note this one also

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 3]
```diff
@@ -8362,7 +8473,11 @@ void ApplicationWindow::clearSelection()
       g->removeMarker();
   }
   else if (m->isA("Note"))
-    dynamic_cast<Note*>(m)->editor()->clear();
+  {
+    auto note = dynamic_cast<Note*>(m);
+    if(note)
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: And this one too

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 4]
```diff
@@ -9565,7 +9758,11 @@ void ApplicationWindow::savedProject()
     foreach(MdiSubWindow *w, folderWindows)
     {
       if (w->isA("Matrix"))
-        (dynamic_cast<Matrix*>(w))->undoStack()->setClean();
+      {
+        Matrix* m = dynamic_cast<Matrix*>(w);
+        if (m)
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: Maybe if(m && m->undoStack()) for better safety? Not sure if we can trust the undo stack pointer in the Matrix class.

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 5]
```diff
@@ -9724,11 +9921,17 @@ void ApplicationWindow::deleteSelectedItems()
...
+      if(!fli)
+        continue;
+      Folder *f = fli->folder();
       if (deleteFolder(f))
         delete item;
-    } else
-      dynamic_cast<WindowListItem*>(item)->window()->close();
+    } else {
+      auto wli = dynamic_cast<WindowListItem*>(item);
+      if(wli)
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: Maybe I'm paranoid after seeing some qti code, but what about `if(wli && wli->window())`?

**Comment by eXeC64** [comment]
Okay, I've resolved the merge conflict - but I've not yet gone through your feedback :)

File: Code/Mantid/MantidPlot/src/ApplicationWindow.cpp [hunk 6]
```diff
@@ -14274,7 +14535,12 @@ void ApplicationWindow::showFolderPopupMenu(Q3ListViewItem *it, const QPoint &p,
   cm.insertItem(tr("&Find..."), this, SLOT(showFindDialogue()));
   cm.insertSeparator();
   cm.insertItem(tr("App&end Project..."), this, SLOT(appendProject()));
-  if (dynamic_cast<FolderListItem*>(it)->folder()->parent())
+
+  auto fli = dynamic_cast<FolderListItem*>(it);
+  if(!fli)
+    return;
+
```
Comment (FedeMPouzols) [inline Code/Mantid/MantidPlot/src/ApplicationWindow.cpp]: Being paranoid again, what about a check here for (!fli->folder())` which is used below a couple of times?

**Comment by FedeMPouzols** [comment]
Good, I've added all my comments now. Take those just as suggestions from a paranoid mind ;)

**Comment by eXeC64** [comment]
Okay, I dug into the classes you highlighted, and the pointers you referred to appear to be created or initialised in the constructors, so they'll always be valid.

It ought be safe to merge this branch without further changes.

**Comment by FedeMPouzols** [comment]
Hum, the merge build failed on osx because of the git checkout issue. False alarm. I think it's safe to merge this in before more conflicts arise.

**Comment by eXeC64** [comment]
Jenkins, retest this please.

**Comment by FedeMPouzols** [comment]
Jenkins, retest this  please

**Comment by eXeC64** [comment]
The only failures now seem to be the flaky `RemoveExpDecay` test. Ought to be safe to merge.

**Comment by FedeMPouzols** [comment]
At last it compiled on all platform and the test failures are absolutely not related to this.

It seems that there's a merge conflict again though.

**Comment by eXeC64** [comment]
This is actually the exact same merge conflict, but in reverse, as a result of #103 being reverted.

**Pull request merged**